pkg/csi/cinder/openstack/openstack_snapshots.go

```diff
@@ -0,0 +1,74 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package openstack.snapshots provides an implementation of Snapshot related
+// cinder functions using Gohpercloud.
+package openstack
+
+import (
+	"github.com/golang/glog"
+	"github.com/gophercloud/gophercloud/openstack/blockstorage/v3/snapshots"
+)
+
+// CreateSnapshot issues a request to take a Snapshot of the specified Volume with the cooresponding ID and
+// returns the resultant gophercloud Snapshot Item upon success
+func (os *OpenStack) CreateSnapshot(name, volID, description string, tags *map[string]string) (*snapshots.Snapshot, error) {
+	opts := &snapshots.CreateOpts{
+		VolumeID:    volID,
+		Name:        name,
+		Description: description,
+	}
+	if tags != nil {
+		opts.Metadata = *tags
+	}
+
+	snap, err := snapshots.Create(os.blockstorage, opts).Extract()
+	if err != nil {
+		return &snapshots.Snapshot{}, err
+	}
+	// There's little value in rewrapping these gophercloud types into yet another abstraction/type, instead just
+	// return the gohpercloud item
+	return snap, nil
+}
+
+// ListSnapshots retrieves a list of active snapshots from Cinder for the cooresponding Tenant.  We also
+// provide the ability to provide limit and offset, however for accurate pagination suggest this is handled
+// instead by the caller.  In addition the filters argument provides a mechanism for passing in valid
+// filter strings to the list operation.  Valid filter keys are:  Name, Status, VolumeID (TenantID has no
+// effect)
+func (os *OpenStack) ListSnapshots(limit, offset int, filters map[string]string) ([]snapshots.Snapshot, error) {
+	opts := snapshots.ListOpts{}
+	pages, err := snapshots.List(os.blockstorage, opts).AllPages()
+	if err != nil {
+		glog.V(3).Infof("Failed call to retrieve snapshots from Cinder: %v", err)
+		return nil, err
+	}
+	snaps, err := snapshots.ExtractSnapshots(pages)
+	if err != nil {
+		glog.V(3).Infof("Failed to extract snapshot pages from Cinder: %v", err)
+		return nil, err
+	}
+	// There's little value in rewrapping these gophercloud types into yet another abstraction/type, instead just
+	// return the gohpercloud item
+	return snaps, nil
+
+}
+
+// DeleteSnapshot issues a request to delete the Snapshot with the specified ID from the Cinder backend
+func (os *OpenStack) DeleteSnapshot(snapID string) error {
+	err := snapshots.Delete(os.blockstorage, snapID).ExtractErr()
```

Comment:
You dont print any glog message in case of error ! This will silently fail

Reply:
Well, it wouldn't be a silent failure really, the caller would receive the error and I would expect the error to be dealt with.  I get your point though and I'll add a log message.  Thanks!

Reply:
Like I commented to another error: called can take also some responsibility of this? We could log that also in caller function